Support free threaded Python versions like '3.13t'

colesbury opened the pull request:
Python wheels, pyenv, and a number of other tools use 't' in the Python version number to identify free threaded builds. For example, '3.13t', '3.14.0a1', '3.14t-dev'.

This PR supports that syntax in `actions/setup-python`, strips the "t", and adds "-freethreading" to the architecture to select the correct Python version.

See #771 and https://github.com/actions/python-versions/pull/319

renxida commented:
Meanwhile, if you can't wait, here at https://github.com/nod-ai/shark-ai/actions/workflows/ci_linux_x64_nogil-libshortfin.yml

we are using

```yaml
    - name: Setup Python
      uses: deadsnakes/action@e640ac8743173a67cca4d7d77cd837e514bf98e8 # v3.2.0
      with:
        python-version: "3.13-dev"
        nogil : true
```

rgommers commented:
If anyone needs the functionality in this PR, I'd recommend to use https://github.com/Quansight-Labs/setup-python. That is a fork of this action with the sole purpose of providing the functionality in this PR (free-threading support) until this PR gets merged.

Data point for reviewers: we've exercised this PR quite heavily by now in projects like NumPy and SciPy, on Linux, macOS and Windows (x86-64 and aarch64/arm64), and it's been working smoothly to date.

priyagupta108 commented:
Hi @colesbury 👋,
Thank you for your valuable contribution!
During our testing, we observed that the changes in this PR do not support specifying the patch version (e.g., 3.13.1t) for the `python-version` input for the `free-threaded` binary. Please ensure that the `python-version` input values align with the existing [input syntax](https://github.com/actions/setup-python?tab=readme-ov-file#supported-version-syntax). 
Additionally, To keep the documentation up-to-date, please update this PR to incorporate the corresponding documentation updates.

Thank you!

ddelange commented:
Also noting here that 3.13 and 3.13t currently don't live happily next to each other on Windows: https://github.com/Quansight-Labs/setup-python/issues/5#issuecomment-2586307554

colesbury commented:
Hi @priyagupta108, thank you for reviewing the PR. I've fixed the handling of versions like `3.13.1t` and added a test for that. I've also updated the README and "advanced-usage" documentation.

priyagupta108 commented:
Hi @colesbury 👋,

Thank you for making the updates! During our testing of this PR, we've identified a few issues related to free-threaded Python version inputs:

* **Prerelease Versions**: The current changes do not support `semver` notation for prerelease versions such as alpha, beta, and RC. For example, versions like 3.14.0-alpha.1, 3.9.0-beta.2, and 3.9.0-rc.1 should be supported. 
Additionally, versions such as 3.14.0a1t and 3.14.0ta1 are mentioned in the documentation changes, but they are failing to run as shown in the screenshot below. Please note that the versions 3.14.0a1t and 3.14.0ta1 do not appear to be valid semver notation according to the [semver guidelines](https://github.com/npm/node-semver?tab=readme-ov-file#prerelease-tags).<br>
   <img width="1025" alt="Screenshot 2025-01-29 at 11 38 26 AM" src="https://github.com/user-attachments/assets/a387de5b-b965-4705-959b-31f22b584db3" />
  <img width="1033" alt="Screenshot 2025-01-29 at 11 38 55 AM" src="https://github.com/user-attachments/assets/fc4bea1c-3341-4840-939b-d3eb075cced8" />

* **Version Ranges**: Various range syntaxes (e.g., hyphen ranges, x-ranges) are unsupported.
* **python-version-file Input**: Unable to read the Python version from `pyproject.toml`. It seems that the `python-version-file` input is not functioning correctly, as shown in the below screenshot.<br>
![Screenshot 2025-01-28 at 6 21 11 PM](https://github.com/user-attachments/assets/add2a0af-2d0c-4362-8e62-37f2a77dd0b3)

Here is a [demo run](https://github.com/setup-actions-demo/sample-python/actions/runs/13025813240) that includes prerelease versions, version ranges and python-version-file input examples. Please ensure that the PR changes also align with this input syntax for free-threaded Python versions.

To ensure full compatibility with all supported Python version formats and inputs including [python-version-file](https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md#using-the-python-version-file-input) input, please address these issues and ensure the implementation fully supports [semantic versioning (semver) guidelines](https://github.com/npm/node-semver) to handle different version formats correctly. Additionally, update the documentation to provide clarity on the syntax for free-threaded Python versions.
For details on supported Python version inputs, please refer the action's [documentation](https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md#using-the-python-version-input).

Thank you!

colesbury commented:
Hi @priyagupta108, thanks for reviewing the PR again.

* I've added an optional "freethreaded" input. As you point out, the "3.13t" syntax is not `semver` notation and there isn't a good way to represent it using semver. Users can specify "freethreading: true" and specify the version using `semver`
* Prerelease notation like "3.14.0a4t" now works. I removed the "3.13.0ta4" notation because that's not supported by other Python tools.
* Version ranges: These now work, but require setting the "freethreaded: true" input to get the freethreaded Python version.
* python-version-file input: This also works and requires setting "freethreaded: true" to get the freethreaded Python version

I've included examples in "advanced-usage.md".

Thank you for the demo run. I've modified it and run it with these changes here:
https://github.com/colesbury/sample-python/actions/runs/13060324198

colesbury commented:
Hi @priyagupta108, I've made an additional change:

* The `outputs.python-version` now includes the "t" suffix for free threaded builds. We've found that the `python-version` output is commonly used for cache keys (see [this github search](https://github.com/search?q=%2Foutputs.python-version%2F&type=code)) and having a different identifier helps avoid problems with mixing caches between incompatible Python builds.

priyagupta108 commented:
Hi @colesbury,
Thank you for the prompt updates. We have reviewed the changes and have a few concerns:

- **Regarding new notation**: The prerelease notation `"3.14.0a4"` introduced does not comply with the Semantic Versioning (SemVer) specification and is not supported by any `setup-actions`. Introducing it here could lead to inconsistencies across all actions. We recommend not including this notation in both the free-threaded and default Python version inputs to ensure that only valid SemVer formats are supported across all `setup-actions`. 

- **Cache-key conflicts**: The `outputs.python-version` correctly reflects the Python version according to the binary type. However, to avoid cache conflicts between free-threaded and default builds when using the `setup-python` cache feature, the cache key should be identical for both. Could you please make the necessary adjustments for the free-threaded cache key? 
Here are the demonstration runs showcasing the caching feature with `setup-python`:
  - **Cache Creation:** [View Details](https://github.com/setup-actions-demo/sample-python/actions/runs/13199184430/job/36847118543#step:9:5)
  - **Cache Restoration:** [View Details](https://github.com/setup-actions-demo/sample-python/actions/runs/13199203638/job/36847177305#step:3:44)

Please feel free to reach out if you need any further clarification or assistance.

ddelange commented:
> * However, to avoid cache conflicts between free-threaded and default builds when using the `setup-python` cache feature, the cache key should be identical for both.

I'd argue that the cache key be *should* be unique in order to avoid conflicts: I don't want to restore a 3.13t venv on a 3.13 runner or vice versa.

Furthermore, for pre-commit related issues, I don't even want to restore a 3.13.0 venv on a 3.13.1 runner (patch updates I get automatically when specifying `3.13` in setup-python, which I do want).

Solution is to simply use the setup-python output `${{ env.pythonLocation }}` in the cache key, which is always unique and always safe.

priyagupta108 commented:
@ddelange 👋, Sorry for the confusion and thanks for pointing it out. Yes, the cache key should be unique. I have edited my previous comment.

priyagupta108 commented on docs/advanced-usage.md:
````diff
@@ -77,6 +77,30 @@ steps:
 - run: python my_script.py
 ```
 
+You can specify the [free threading](https://docs.python.org/3/howto/free-threading-python.html) version of Python by setting the `freethreaded` input to `true` or by using the special **t** suffix in some cases.  Pre-release free threading versions can be specified like `3.14t-dev`.
````
The documentation implies that all pre-release versions (alpha, beta, rc, dev) should be valid when using the "t" suffix. However, the functionality works only for 3.14t-dev. 
Please update the documentation to accurately reflect the current behavior.

priyagupta108 commented:
@colesbury 👋,
Since free-threaded Python is only available starting with the 3.13 release, attempting to use version 3.12 or earlier results in the error message, as shown in the screenshot below:

<img width="1044" alt="Screenshot 2025-02-14 at 3 14 27 PM" src="https://github.com/user-attachments/assets/0388fa83-0078-4665-86fa-46d46679e8e3" /><br>

For better clarity, it would be helpful to update the error message to indicate that free-threaded Python is not supported for versions 3.12 or earlier.
Thank you!

colesbury commented:
Hi @priyagupta108, I've updated the documentation and error message.

priyagupta108 commented:
Hello @colesbury,
Thank you once again for the updates and your continued efforts!
To ensure proper testing of this feature, it would be beneficial to incorporate end-to-end tests for free-threaded Python versions in this PR. The `3.13.1` and `3.14.0-alpha.5` free-threaded versions are now available in the `actions/python-versions` repository, with the remaining versions to be released soon. Please include these two versions in the e2e tests. The relevant test cases can be added to the following workflows:

- [.github/workflows/e2e-tests.yml](https://github.com/actions/setup-python/blob/main/.github/workflows/e2e-tests.yml)
- [.github/workflows/test-python.yml](https://github.com/actions/setup-python/blob/main/.github/workflows/test-python.yml)
- [.github/workflows/e2e-cache.yml](https://github.com/actions/setup-python/blob/main/.github/workflows/e2e-cache.yml)

hugovk commented:
This has now been released as v5.5.0 and v5:

* https://github.com/actions/setup-python/releases/tag/v5.5.0
* https://github.com/actions/setup-python/releases/tag/v5

I wrote up some instructions:

* https://hugovk.dev/blog/2025/free-threaded-python-on-github-actions/
